daemon/rename.go

```diff
@@ -118,10 +119,68 @@ func (daemon *Daemon) ContainerRename(oldName, newName string) (retErr error) {
 			return err
 		}
 
-		err = sb.Rename(strings.TrimPrefix(container.Name, "/"))
-		if err != nil {
+		if err := sb.Rename(newName[1:]); err != nil {
 			return err
 		}
+		defer func() {
+			if retErr != nil {
+				if err := sb.Rename(oldName); err != nil {
+					log.G(context.TODO()).WithFields(log.Fields{
+						"sandboxID": sid,
+						"oldName":   oldName,
+						"newName":   newName,
+						"error":     err,
+					}).Errorf("failed to revert sandbox rename")
+				}
+			}
+		}()
+
+		var updatedEps []struct {
+			ep       *libnetwork.Endpoint
+			epConfig *network.EndpointSettings
+			dnsNames []string
+		}
+		defer func() {
+			if retErr != nil {
+				for _, update := range updatedEps {
+					update.epConfig.DNSNames = update.dnsNames
+					if err := update.ep.UpdateDNSNames(update.epConfig.DNSNames); err != nil {
+						log.G(context.TODO()).WithFields(log.Fields{
+							"sandboxID": sid,
+							"oldName":   oldName,
+							"newName":   newName,
+							"error":     err,
+						}).Errorf("failed to revert DNSNames update")
+					}
+				}
+			}
+		}()
```

Comment:
Instead of managing the undo stack by hand, why not let the compiler handle it all for you by lowering the `defer` inside the loop?